internal/uvm/scsi.go

```diff
@@ -55,40 +100,107 @@ func (uvm *UtilityVM) allocateSCSI(ctx context.Context, hostPath string, uvmPath
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	si := uvm.scsiLocations[controller][lun]
-	if si.hostPath != "" {
+	if si := uvm.scsiLocations[controller][lun]; si != nil {
 		log.G(ctx).WithFields(logrus.Fields{
-			"hostPath":   si.hostPath,
-			"uvmPath":    si.uvmPath,
+			"hostPath":   si.HostPath,
+			"uvmPath":    si.UvmPath,
 			"isLayer":    si.isLayer,
 			"refCount":   si.refCount,
 			"controller": controller,
 			"lun":        lun,
 		}).Debug("removed SCSI location")
-		uvm.scsiLocations[controller][lun] = scsiInfo{}
+		uvm.scsiLocations[controller][lun] = nil
 	}
 }
 
 // Lock must be held when calling this function.
 func (uvm *UtilityVM) findSCSIAttachment(ctx context.Context, findThisHostPath string) (int, int32, string, error) {
 	for controller, luns := range uvm.scsiLocations {
 		for lun, si := range luns {
-			if si.hostPath == findThisHostPath {
-				log.G(ctx).WithFields(logrus.Fields{
-					"hostPath":   si.hostPath,
-					"uvmPath":    si.uvmPath,
-					"isLayer":    si.isLayer,
-					"refCount":   si.refCount,
-					"controller": controller,
-					"lun":        lun,
-				}).Debug("found SCSI location")
-				return controller, int32(lun), si.uvmPath, nil
+			if si != nil {
+				if si.HostPath == findThisHostPath {
+					log.G(ctx).WithFields(logrus.Fields{
+						"hostPath":   si.HostPath,
+						"uvmPath":    si.UvmPath,
+						"isLayer":    si.isLayer,
+						"refCount":   si.refCount,
+						"controller": controller,
+						"lun":        lun,
+					}).Debug("found SCSI location")
+					return controller, int32(lun), si.UvmPath, nil
+				}
 			}
 		}
 	}
 	return -1, -1, "", ErrNotAttached
 }
 
+// RemoveSCSI removes a SCSI disk from a utility VM.
+func (uvm *UtilityVM) RemoveSCSI(ctx context.Context, hostPath string) error {
+	uvm.m.Lock()
+	defer uvm.m.Unlock()
+
+	if uvm.scsiControllerCount == 0 {
+		return ErrNoSCSIControllers
+	}
+
+	// Make sure is actually attached
+	controller, lun, uvmPath, err := uvm.findSCSIAttachment(ctx, hostPath)
+	if err != nil {
+		return err
+	}
+
+	sm := uvm.scsiLocations[controller][lun]
+	sm.refCount--
+	if sm.refCount > 0 {
+		return nil
+	}
+
+	scsiModification := &hcsschema.ModifySettingRequest{
+		RequestType:  requesttype.Remove,
+		ResourcePath: fmt.Sprintf(scsiResourceFormat, strconv.Itoa(controller), lun),
+	}
+
+	// Include the GuestRequest so that the GCS ejects the disk cleanly if the
+	// disk was attached/mounted
+	//
+	// Note: We always send a guest eject even if there is no UVM path in lcow
+	// so that we synchronize the guest state. This seems to always avoid SCSI
+	// related errors if this index quickly reused by another container.
+	if uvm.operatingSystem == "windows" && uvmPath != "" {
+		scsiModification.GuestRequest = guestrequest.GuestRequest{
+			ResourceType: guestrequest.ResourceTypeMappedVirtualDisk,
+			RequestType:  requesttype.Remove,
+			Settings: guestrequest.WCOWMappedVirtualDisk{
```

Comment:
The if/else only needs to provide WCOWMappedVirtualDisk vs LCOWMappedVirtualDisk. The GuestRequest is otherwise the same. Can this be factored out?

Reply:
Just to clarify you mean just make the guestrequest above and in this if/else just add the Settings (LCOW vs WCOW) portion to it?

Reply:
Oh that could work fine. I had been thinking the opposite, have a var settings {}interface above the if/else and set it, then use this to construct the GuestRequest object. Either way, let's just have less duplicate code.

(Again, I'm aware that this is existing code :))